src/expr.cc

```diff
@@ -546,30 +546,39 @@ FunctionCall::FunctionCall(Expression *expr, const AssignmentList &args, const L
 */
 void FunctionCall::prepareTailCallContext(const std::shared_ptr<Context> context, std::shared_ptr<Context> tailCallContext, const AssignmentList &definition_arguments)
 {
-	if (this->resolvedArguments.empty()) {
+	if (this->resolvedArguments.empty() && !(definition_arguments.empty() && this->arguments.empty())) {
 		// Figure out parameter names
 		ContextHandle<EvalContext> ec{Context::create<EvalContext>(context, this->arguments, this->loc)};
 		this->resolvedArguments = ec->resolveArguments(definition_arguments, {}, false);
-		// Assign default values for unspecified parameters
-		for (const auto &arg : definition_arguments) {
-			if (this->resolvedArguments.find(arg.name) == this->resolvedArguments.end()) {
-				this->defaultArguments.emplace_back(arg.name, arg.expr ? arg.expr->evaluate(context) : ValuePtr::undefined);
-			}
-		}
 	}
 
+	/*
+	for (const auto &arg :this->resolvedArguments) {
+		tailCallContext->set_variable(arg.second.name,
+			arg.second.expr ?
+				(arg.first ? arg.second.expr->evaluate(context->getParent()) : arg.second.expr->evaluate(context)) :
+				ValuePtr::undefined);
+	}
+	//*/
+
+	//*
 	std::vector<std::pair<std::string, ValuePtr>> variables;
-	variables.reserve(this->defaultArguments.size() + this->resolvedArguments.size());
+	variables.reserve(this->resolvedArguments.size());
 	// Set default values for unspecified parameters
-	variables.insert(variables.begin(), this->defaultArguments.begin(), this->defaultArguments.end());
 	// Set the given parameters
-	for (const auto &ass : this->resolvedArguments) {
-		variables.emplace_back(ass.first, ass.second->evaluate(context));
+	for (const auto &arg : this->resolvedArguments) {
+		variables.emplace_back(arg.second.name,
+				arg.second.expr ?
+				(arg.first ? arg.second.expr->evaluate(context->getParent()) : arg.second.expr->evaluate(context)) :
```

Comment:
This is one of the parts I'm still unsure if its 100% correct.

`arg.first` is a bool where true indicates that the assignment is defaulted, so it should be evaluated in the context of the function *definition* rather than the function call.  So the code evaluates from `context->getParent()`, which is the same way that `Context::setVariables` handles default arguments(both before and after my changes). 

I'm just not sure if the parent context is really correct or sufficient in the case of some recursive function where the function itself could be re-defined within its own `let()` expressions.

Reply:
Test case from IRC:
`foo = function() let(foo = function() 5) foo(); //infinite recursion`